# **Credit Card Fraud Detection**

Import all necessary liabraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score, confusion_matrix, accuracy_score

from pylab import rcParams
rcParams['figure.figsize'] = 8,8

In [2]:
data = pd.read_csv('creditcard.csv')

In [3]:
data.head()     # data has been encoded or masked

,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,...,V21,V22,V23,V24,V25,V26,V27,V28,V29,Target
0,0.114697,0.796303,-0.149553,-0.823011,0.878763,-0.553152,0.939259,-0.108502,0.111137,-0.390521,...,-0.335776,-0.807853,-0.055940,-1.025281,-0.369557,0.204653,0.242724,0.085713,0.89,0
1,-0.039318,0.495784,-0.810884,0.546693,1.986257,4.386342,-1.344891,-1.743736,-0.563103,-0.616315,...,-1.377003,-0.072200,-0.197573,1.014807,1.011293,-0.167684,0.113136,0.256836,85.00,0
2,2.275706,-1.531508,-1.021969,-1.602152,-1.220329,-0.462376,-1.196485,-0.147058,-0.950224,1.560463,...,-0.193271,-0.103533,0.150945,-0.811083,-0.197913,-0.128446,0.014197,-0.051289,42.70,0
3,1.940137,-0.357671,-1.210551,0.382523,0.050823,-0.171322,-0.109124,-0.002115,0.869258,-0.001965,...,0.157994,0.650355,0.034206,0.739535,0.223605,-0.195509,-0.012791,-0.056841,29.99,0
4,1.081395,-0.502615,1.075887,-0.543359,-1.472946,-1.065484,-0.443231,-0.143374,1.659826,-1.131238,...,0.224157,0.821209,-0.137223,0.986259,0.563228,-0.574206,0.089673,0.052036,68.00,0


In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56962 entries, 0 to 56961
Data columns (total 30 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   V1      56962 non-null  float64
 1   V2      56962 non-null  float64
 2   V3      56962 non-null  float64
 3   V4      56962 non-null  float64
 4   V5      56962 non-null  float64
 5   V6      56962 non-null  float64
 6   V7      56962 non-null  float64
 7   V8      56962 non-null  float64
 8   V9      56962 non-null  float64
 9   V10     56962 non-null  float64
 10  V11     56962 non-null  float64
 11  V12     56962 non-null  float64
 12  V13     56962 non-null  float64
 13  V14     56962 non-null  float64
 14  V15     56962 non-null  float64
 15  V16     56962 non-null  float64
 16  V17     56962 non-null  float64
 17  V18     56962 non-null  float64
 18  V19     56962 non-null  float64
 19  V20     56962 non-null  float64
 20  V21     56962 non-null  float64
 21  V22     56962 non-null  float64
 22

In [5]:
data.describe()

,V1,V2,V3,V4,V5,V6,V7,V8,V9,V10,...,V21,V22,V23,V24,V25,V26,V27,V28,V29,Target
count,56962.000000,56962.000000,56962.000000,56962.000000,56962.000000,56962.000000,56962.000000,56962.000000,56962.000000,56962.000000,...,56962.000000,56962.000000,56962.000000,56962.000000,56962.000000,56962.000000,56962.000000,56962.000000,56962.000000,56962.000000
mean,0.013286,0.006610,-0.004263,0.001496,-0.003508,-0.003079,0.000141,-0.006501,0.001564,0.003176,...,-0.002253,-0.004938,0.004008,-0.001017,-0.000872,0.004513,0.001385,-0.001990,87.658797,0.001720
std,1.941231,1.611224,1.516853,1.419107,1.431731,1.359647,1.301800,1.201138,1.103688,1.095156,...,0.735871,0.730301,0.621623,0.604191,0.521003,0.481857,0.416750,0.325640,258.042879,0.041443
min,-36.510583,-47.429676,-48.325589,-5.560118,-113.743307,-20.054615,-28.215112,-41.484823,-9.481456,-20.949192,...,-21.620120,-8.593642,-32.828995,-2.822684,-8.696627,-1.778061,-8.878665,-15.430084,0.000000,0.000000
25%,-0.908591,-0.600208,-0.893961,-0.847617,-0.694059,-0.770145,-0.551096,-0.211190,-0.642072,-0.535121,...,-0.230210,-0.549030,-0.160179,-0.356599,-0.315738,-0.322766,-0.070204,-0.052351,5.615000,0.000000
50%,0.036251,0.061840,0.170910,-0.022094,-0.051828,-0.278939,0.039155,0.019903,-0.052607,-0.095676,...,-0.032379,-0.000329,-0.009334,0.040045,0.018074,-0.047859,0.001451,0.011354,21.900000,0.000000
75%,1.319727,0.801163,1.016897,0.738688,0.615454,0.392801,0.569769,0.324905,0.599634,0.453059,...,0.184809,0.524484,0.149985,0.437657,0.348895,0.245286,0.090737,0.077483,77.500000,0.000000
max,2.446505,18.902453,3.985446,15.304184,29.016124,73.301626,120.589494,18.282168,10.370658,15.236028,...,22.614889,6.790452,20.803344,4.584549,7.519589,3.517346,31.612198,22.620072,25691.160000,1.000000


In [6]:
data.shape      #shape of the dataset

(56962, 30)

**We have large enough dataset to apply Random Forest or any other good ML algorithm**

In [7]:
# Let's check if data is balanced or not

data['Target'].value_counts()

Target
0    56864
1       98
Name: count, dtype: int64

**We have highly imbalanced data having just 98 fraudulent transactions and 56864 legitimate trsactions**

In [8]:
data['Target'].value_counts(normalize=True)

Target
0    0.99828
1    0.00172
Name: proportion, dtype: float64

#### Things to do when we have imbalanced data:
1. Oversampling (SMOTE Technique)
2. Undersampling(Loss of Data)
3. Use Tree Based models only (Decision Tree, Random Forest, XGBoost, Adaboost, GradientBoost).

In [9]:
# Now let's divide the input and output data

X = data.drop('Target', axis=1)
y = data['Target']

* Split the dataset into train and test using stratified sampling on our dependent variable. 
* Using a stratified sampling ensures the distribution of dependent variable remains same across train and test datasets

In [10]:
# Split the dataset into Training and Testing

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify= y, random_state=42)

The stratify parameter asks whether you want to retain the same proportion of classes in the train and test sets that are found in the entire original dataset. It means if there are 60% 'yes' class data and 40% 'no' class data in the full dataset, then in train and test data split this proportion should be maintained.

## Apply ML to the Data

In [11]:
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

In [12]:
param_grid = { 
    'n_estimators': [25, 50, 100, 150], 
    'max_depth': [1,2,3,4,5,6,7,8,9], 
    'max_leaf_nodes': [3,6,9], 
}

In [13]:
grid_search = GridSearchCV(RandomForestClassifier(), param_grid=param_grid) 
grid_search.fit(X_train, y_train)
print(grid_search.best_estimator_)

RandomForestClassifier(max_depth=6, max_leaf_nodes=9, n_estimators=150)


In [16]:
random_search = RandomizedSearchCV(RandomForestClassifier(), param_distributions=param_grid)
random_search.fit(X_train, y_train)
print(random_search.best_estimator_)

RandomForestClassifier(max_depth=5, max_leaf_nodes=9, n_estimators=25)


### Let's build a model with max_depth=5, max_leaf_nodes=9, n_estimators=25

In [17]:
rf_classifier_1 = RandomForestClassifier(n_estimators=25, max_depth=5, max_leaf_nodes=9)
rf_classifier_1.fit(X_train, y_train)

RandomForestClassifier(max_depth=5, max_leaf_nodes=9, n_estimators=25)

**n_estimator** is the number of trees and ideally the value of this paramater `should not be less than 100` and **max_depth** helps us control the length of each tree in Random Forest

In [19]:
y_pred = rf_classifier_1.predict(X_test)

In [20]:
accuracy_score(y_test, y_pred)

0.9991222680593347

In [23]:
y_pred_train = rf_classifier_1.predict(X_train)

In [24]:
accuracy_score(y_train, y_pred_train)

0.9993636024490333

**Accuracy is not a good metric to evaluate our model since the dataset is heavily imbalanced. Thus we will use roc_auc score and confusion matrix**

In [25]:
roc_auc_score(y_test, y_pred)

0.7749560362261496

In [26]:
roc_auc_score(y_train, y_pred_train)

0.8141025641025641

### Let's build a model with max_depth=6, max_leaf_nodes=9, n_estimators=150

In [27]:
rf_classifier_2 = RandomForestClassifier(max_depth=6, max_leaf_nodes=9, n_estimators=150)
rf_classifier_2.fit(X_train, y_train)

RandomForestClassifier(max_depth=6, max_leaf_nodes=9, n_estimators=150)

In [28]:
y_pred_1 = rf_classifier_2.predict(X_test)

In [29]:
accuracy_score(y_test, y_pred_1)

0.9990344948652682

In [30]:
y_pred_train1 = rf_classifier_2.predict(X_train)

In [31]:
accuracy_score(y_train, y_pred_train1)

0.9993855471921701

In [32]:
roc_auc_score(y_test, y_pred_1)

0.7749120724522992

In [33]:
roc_auc_score(y_train, y_pred_train1)

0.8269120857380072

**With both set of hyperparameters model is seems to be working almost same.**